### ISSUE TITLE
frida-ps -U   failed to open USB device: Input/Output Error

### ISSUE THREAD

**Issue by WyntersN** [frida/frida#3278]
![image](https://github.com/user-attachments/assets/69c3bc6d-a673-4385-acd9-9843519867d8)

Why?

**Comment by Malarie2** [comment]
same

**Comment by Malarie2** [comment]
did u find solution?

**Comment by SouthWind0** [comment]
same

**Comment by lufei0099** [comment]
Lower the version ... eg :16.0.0

**Comment by zonedark8** [comment]
same

**Comment by dlnn** [comment]
Lower to 16.5.2
 https://frida.re/news/2024/10/10/frida-16-5-3/
Many about usb errors on 16.5.3 or later

**Comment by jpstotz** [comment]
I assume this happens because frida because of a bug now always has two USB devices internally: 

1. A dummy device without actual name and id.
2. The actual USB device

If you only execute `frida-ps -U` the first device (thus the dummy device)  is used which then fails.

As mentioned in https://github.com/frida/frida/issues/3283 you can specify the device id as workaround.

**Comment by grekland** [comment]
**Running following command will let you understand why it happens:**
`devcon listclass USBDevice`

### C:\ > devcon listclass USBDevice
```
Listing 3 devices in setup class "USBDevice" 
USB\XXXXXXXXXXX              : Apple Mobile Device USB Composite Device             (1st item)
USB\YYYYYYYYYYYY               : Apple Mobile Device USB Device                     (2nd item)
```

Frida does the same thing to get its USB devices. The error you are getting arises due to the order of which the devices are listed. As you can see, **Apple Mobile Device USB Device** (the _"actual"_ device), is listed **after** the composite one _(aka dummy device)_. Therefore, when you use the **-U** header and Frida gets the **first listed item**, the program will raise an error because it's gotten the composite device and not the "real" one.


### Solutions
Easiest way to get around this is replacing the **-U** header with the **-D** header. The **-D** header takes your device's ID (see frida-ls-devices). Doing this gets quite tedious, though. 

However, because the device ID is static and does not change, it is also possible to create a **.bat** file to handle the ID header for us. For example (test-frida.bat): 
```
@ECHO OFF
rem parse either: -F (attach-frontmost), -n (attach-name), -p (attach-pid), -N (attach-identifier)
SET _id=your-device-id

IF [%1]==[] (
    GOTO :error
)

IF /I "%1"=="-F" (
    SET _method=NUL
) ELSE (
    SET _method=%2
)

frida %1 %_method% -D %_id%
GOTO :EOF

:error
ECHO Possible ArgError: Missing or incorrect arguments.
```
You can then add this .bat file to the environment path.

Another alternative solution is setting the ID as an environment variable. This will require us to type some annoying symbols, but at least we won't need to copy and paste our ID every we want to attach Frida to an app. Example:
`frida -D $Env:ID -F`

**Comment by jpstotz** [comment]
@grekland Your analysis of the problem, that the compound devices is the dummy device sounds reasonable. In the end the best way would be if you could make Frida to ignore/filter out the compound device. If I am not mistaken USB compound devices usually don't offer own services, they just allow one physical USB device to be split into multiple logical devices. Therefore filtering them out should have no negative effect on Frida.

The only question that remains is in which subproject of frida is the code hidden that is responsible for enumerating the USB devices on each platform? Some time s I really would like to have a "frida code map" where all components and responsibilities (and where they are executed) are mapped so in case of errors back tracking would become easier...

**Comment by grekland** [comment]
@jpstotz I know right. The creator of Frida seems to be quite fond of hiding his code away. pretty ironic  but i respect it lol. 
I like your idea about filtering out the compound. Let's try that!

### Decompiling frida.exe
When we call `frida` in a terminal, we are calling the executable `frida.exe` , located in `C:\Users\User\AppData\Local\Programs\Python\Python311\Scripts`. This poses some difficulties. In order for us to patch or modify `frida.exe`, we would need to carry out a static and/or possibly dynamic analysis of its binary. I don't have enough experience for that, and even if someone in here did, I would imagine it's a rather tiresome task.

> tldr: tedious and difficult

### Alternative approach?
Moreover, in python, we can call enumerate_devices:
```
import frida

frida.enumerate_devices()
```
Let's trace where it comes from. 

1. `pip show frida` tell us that the frida module can be found in: `...\Python\Python311\Lib\site-packages`
2. In the frida module we find `__init__.py`. This file contains following: `get_device_manager().enumerate_devices()`
3. The device manager is defined as: `get_device_manager = core.get_device_manager`
4. _fast forwarding..._
5.  The logic for getting the USB devices can be found in the `_frida` module, specifically under `_frida.DeviceManager`.

The mentioned _frida module is a reference to a file `_frida.pyd`. .pyd files are essentially DLLs, shared libraries. ... compiled python code consisting of only binary code.

### It is what it is
Now that we've been left with no other choice but to decompile the binary, where do we start?

**I found this:**
- [https://github.com/extremecoders-re/pyinstxtractor](url)
- [https://github.com/rocky/python-decompile3](url)

I know some Ghidra. Maybe that could be of use? All of this made me hungry. I'll go get something to eat brb chat

**Comment by Xiao1an** [comment]
Hello, a easy solution

use

frida-ps -D [device ID]

My Device ID?
use frida-ls-devices

**Comment by KDR9666** [comment]
Please help on below issue,

![Image](https://github.com/user-attachments/assets/d8c9a2cd-d272-45b3-a4d0-6259ee76f4a5)

**Comment by jpstotz** [comment]
@WyntersN  Should be fixed since 16.7.1
